pulsar-transaction/buffer/src/main/java/org/apache/pulsar/transaction/buffer/impl/TransactionCursorImpl.java

```diff
@@ -123,4 +204,377 @@ private void addTxnToCommittedIndex(TxnID txnID, long committedAtLedgerId) {
+                    this.status = SnapshotStatus.END;
+                    break;
+            }
+            this.meta = DataFormat.recoverMeta(txn.getTxnMeta());
+            this.position = DataFormat.recoverPosition(txn.getPosition());
+        }
+
+    }
+
+    private CompletableFuture<Void> recoverFromStart(Position currentPosition) {
+        return readPrevEntry(currentPosition)
+                   .thenApply(entry -> new PersistentTxnIndexSnapshot(entry.getData()))
+                   .thenCompose(this::recoverFromEnd);
+
+    }
+
+    private CompletableFuture<Entry> readPrevEntry(Position position) {
+        PositionImpl currentPos = (PositionImpl) position;
+        if (currentPos.getEntryId() == 0) {
+            return FutureUtil.failedFuture(
+                new TransactionIndexRecoveringError("Not found the prev position of the current position " + position));
+        }
+
+        PositionImpl prevPosition = PositionImpl.get(currentPos.getLedgerId(), currentPos.getEntryId() - 1);
+        return readSpecifiedPosEntry(prevPosition);
+    }
+
+    private CompletableFuture<Void> recoverFromMiddle(PersistentTxnIndexSnapshot snapshot) {
+        return readSpecifiedPosEntry(snapshot.position)
+                   .thenApply(entry -> new PersistentTxnIndexSnapshot(entry.getData()))
+                   .thenCompose(startBlock -> recoverFromStart(snapshot.position));
+
+    }
+
+    private CompletableFuture<Entry> readSpecifiedPosEntry(Position position) {
+        CompletableFuture<Entry> readFuture = new CompletableFuture<>();
+
+        PositionImpl readPos = (PositionImpl) position;
+        LedgerHandle ledger = indexCursor.get().getCurrentCursorLedger();
+
+        ledger.asyncReadEntries(readPos.getEntryId(), readPos.getEntryId(), (rc, handle, entries, ctx) -> {
+            if (rc != BKException.Code.OK) {
+                readFuture.completeExceptionally(BKException.create(rc));
+            } else {
+                if (entries.hasMoreElements()) {
+                    LedgerEntry ledgerEntry = entries.nextElement();
+                    EntryImpl entry = EntryImpl.create(ledgerEntry.getLedgerId(), ledgerEntry.getEntryId(),
+                                                       ledgerEntry.getEntry());
+
+                    readFuture.complete(entry);
+                } else {
+                    readFuture.completeExceptionally(new NoSuchElementException(
+                        "No such entry " + readPos.getEntryId() + " in ledger " + handle.getId()));
+                }
+            }
+        }, null);
+
+        return readFuture;
+    }
+
+    private CompletableFuture<Void> recoverFromEnd(PersistentTxnIndexSnapshot snapshot) {
+        return recoverFromLedger(snapshot);
+    }
+
+    private CompletableFuture<Void> recoverFromLedger(PersistentTxnIndexSnapshot snapshot) {
+        return readEntryFromCursorLedger(snapshot.position, indexCursor.get().getCurrentCursorLedger())
+            .thenApply(entries ->
```

Comment:
line 410-417 should be in one function. you shouldn't read all the entries to a list and process the list again. you should just read one entry and apply it (rebuild) to the index.